components/script/dom/window.rs

```diff
@@ -1278,9 +1399,23 @@ impl Window {
 
         // Clean up any active promises
         // https://github.com/servo/servo/issues/15318
-        if let Some(custom_elements) = self.custom_element_registry.get() {
-            custom_elements.teardown();
-        }
+        let document = match &*self.current_state.borrow() {
+            WindowState::Zombie(_) => {
+                unreachable!("Window should be alive when clear_js_runtime is called")
+            },
+            WindowState::Alive(data) => {
+                if let Some(custom_elements) = data.custom_element_registry.get() {
+                    custom_elements.teardown();
+                }
+                data.document
+                    .take()
+                    .expect("Window to have a document when the js runtime is cleared")
+            },
+        };
+
+        // Drop all DOM related data, keeping only the document.
+        *self.current_state.borrow_mut() =
+            WindowState::Zombie(MutNullableDom::new(Some(&*document)));
```

Comment:
Trying to see what happens if I move this call to before the call to `Gc` below..

Reply:
The crash with `/FileAPI/url/url-lifetime.html` was due to the call to `Gc` happening when document has been "taken" above, and not set on the zombie state yet, so either we don't take it above and do another borrow of the state after the Gc, or we set it to zombie(with the document), before doing a `Gc`.

Reply:
The advantage of doing it before the `Gc`, is that if we set the zombie state then, the call to `Gc` potentially collects more than if done before setting the state to zombie(which means a bunch of JS objects are kept alive by rust during the Gc call).